bin/PrintHighestContainingSubstitution/Main.cpp

```diff
@@ -0,0 +1,171 @@
+#include <fcntl.h>
+#include <pasta/AST/AST.h>
+#include <pasta/AST/Macro.h>
+#include <pasta/Compile/Command.h>
+#include <pasta/Compile/Compiler.h>
+#include <pasta/Compile/Job.h>
+#include <pasta/Util/ArgumentVector.h>
+#include <pasta/Util/File.h>
+#include <pasta/Util/FileSystem.h>
+#include <pasta/Util/Init.h>
+#include <unistd.h>
+
+#include <csignal>
+#include <cstdlib>
+#include <fstream>
+#include <iostream>
+#include <memory>
+#include <sstream>
+#include <string>
+#include <unordered_set>
+
+void PrintTokensTo(std::iostream &s, const pasta::TokenRange &tokens) {
+  int i = 0;
+  for (auto &&token : tokens) {
+    if (token.Role() == pasta::TokenRole::kFinalMacroExpansionToken ||
+        token.Role() == pasta::TokenRole::kFileToken) {
+      if (i > 0) {
+        s << ' ';
+      }
+      s << token.Data();
+      i++;
+    }
+  }
+}
+
+class StmtPrintHighestContainingSubstitutionFinder final : public pasta::StmtVisitor {
+public:
+  std::ostream &os;
+
+  virtual ~StmtPrintHighestContainingSubstitutionFinder(void) = default;
+
+  explicit StmtPrintHighestContainingSubstitutionFinder(std::ostream &os_) : os(os_) {}
+
+  void VisitStmt(const pasta::Stmt &stmt) {
+    if (stmt.Kind() != pasta::StmtKind::kImplicitCastExpr
+        && stmt.Kind() != pasta::StmtKind::kImplicitValueInitExpr)
+    {
+      if (const auto sub = stmt.HighestContainingSubstitution())
+      {
+        std::stringstream ss;
+        PrintTokensTo(ss, stmt.Tokens());
+        ss << " is contained in ";
+        assert(sub->BeginToken().has_value());
+        const auto name = sub->BeginToken()->Data();
+        ss << name << " at the highest level";
+        std::cout << ss.str() << "\n";
+      }
+    }
+    for (const auto &child : stmt.Children())
+    {
+      Accept(child);
+    }
+  }
+};
+
+class DeclPrintHighestContainingSubstitutionFinder final : public pasta::DeclVisitor {
+public:
+  std::ostream &os;
+
+  virtual ~DeclPrintHighestContainingSubstitutionFinder(void) = default;
+
+  explicit DeclPrintHighestContainingSubstitutionFinder(std::ostream &os_) : os(os_) {}
+
+  void VisitDeclContext(const pasta::DeclContext &dc) {
+    for (const auto &decl : dc.AlreadyLoadedDeclarations()) {
+      Accept(decl);
+    }
+  }
+
+  void VisitTranslationUnitDecl(const pasta::TranslationUnitDecl &decl) final {
+    VisitDeclContext(decl);
+  }
+
+  void VisitVarDecl(const pasta::VarDecl &decl) final {
+    if (!decl.IsLocalVariableDeclaration()) {
+      const auto &initializer = decl.Initializer();
```

Comment:
I think this returns an optional. Assigning an optional to a reference seems... unusual / sketchy. In practice, most PASTA methods return values, so just assign to values, i.e. `auto`. Often no real value to the `const` here, unless you are indeed dealing with a reference.

Also, you can do: `if (auto init = decl.Initializer()) {` :-D

Reply:
I didn't start writing C++ until about a year ago, so sometimes all the qualifiers and references confuse me :sweat_smile: but I will keep that in mind about references and optionals. I prefer to default to `const` when possible, but maybe that's unconventional as well. I am aware that you can declare variables inside if conditions - for some reason I just forget it in this case :grimacing: